src/TypeResolver.php

```diff
@@ -103,13 +112,123 @@ public function resolve($type, Context $context = null)
+                        'Unexpected type separator'
+                    );
+                }
+                $tokens->next();
+
+            } else if ($token == '?') {
+                if ($parserContext !== self::PARSER_IN_COMPOUND
+                    && $parserContext !== self::PARSER_IN_ARRAY_EXPRESSION) {
+                    throw new \RuntimeException(
+                        'Unexpected nullable character'
+                    );
+                }
+
+                $tokens->next();
+                $type = $this->parseTypes($tokens, $context, self::PARSER_IN_NULLABLE);
+                $types[] = new Nullable($type);
+
+            } else if ($token === '(') {
+                $tokens->next();
+                $type = $this->parseTypes($tokens, $context, self::PARSER_IN_ARRAY_EXPRESSION);
+
+                $resolvedType = new Array_($type);
+
+                // we generates arrays corresponding to the number of '[]'
+                // after the ')'
+                $numberOfArrays = (strlen($tokens->current()) -1) / 2;
+                for ($i = 0; $i < $numberOfArrays - 1; $i++) {
+                    $resolvedType = new Array_($resolvedType);
+                }
+                $types[] = $resolvedType;
+                $tokens->next();
+
+            } else if ($parserContext === self::PARSER_IN_ARRAY_EXPRESSION
+                       && $token[0] === ')'
+                ) {
+                break;
+
+            } else {
+                $type = $this->resolveSingleType($token, $context);
+                $tokens->next();
+                if ($parserContext === self::PARSER_IN_NULLABLE) {
+                    return $type;
+                }
+                $types[] = $type;
+            }
+        }
+
+        if ($token == '|') {
+            throw new \RuntimeException(
+                'A type is missing after a type separator'
+            );
+        }
+        if (count($types) == 0) {
+            if ($parserContext == self::PARSER_IN_NULLABLE) {
+                throw new \RuntimeException(
+                    'A type is missing after a nullable character'
+                );
+            }
+            if ($parserContext == self::PARSER_IN_ARRAY_EXPRESSION) {
+                throw new \RuntimeException(
+                    'A type is missing in an array expression'
+                );
+            }
+            throw new \RuntimeException(
+                'No types in a compound list'
+            );
+        } else if (count($types) == 1) {
+            return $types[0];
+        }
+        return new Compound($types);
+    }
+
+    /**
+     * resolve the given type into a type object
+     *
+     * @param string    $type      the type string, representing a single type
+     * @param Context   $context
+     * @return Type|Array_|Object_
+     */
+    private function resolveSingleType($type, Context $context)
+    {
         switch (true) {
-            case $this->isNullableType($type):
```

Comment:
Note that resolveSingleType(), as its name indicates, checks only a type name (a keyword type, a class name...). However, `?` is not a type, it is an operator, like `(`, `)` or `|`.  Line 116, I separate operators from other things, so I can parse the expression. `[` and `]` could be processed as operators, but I would not add more complexity.